File: credit.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef enum CardType {AMEX, MASTERCARD, VISA, INVALID} CardType;
const char *CardTypeToString[] = {"AMEX", "MASTERCARD", "VISA", "INVALID"};

long get_long() {
    const int BUFFER_SIZE = 80;
    char buffer[BUFFER_SIZE];
    char *ret, *tail;
    long n;
    do {
        printf("Number: ");
        ret = fgets(buffer, BUFFER_SIZE, stdin);
        n = strtol(buffer, &tail, 10);
    } while (n < 0 || ret == NULL || *tail != 0x0a);
    return n;
}

bool is_valid_number(char *n_str) {  // Use Luhn's algorithm.
    int sum = 0;
    for (int i = strlen(n_str) - 2; i >= 0; i-= 2) {
        int product = 2 * (n_str[i] - '0');
        sum += product >= 10 ? 1 + (product % 10) : product;
    }
    for (int i = strlen(n_str) - 1; i >= 0; i-= 2) {
        sum += n_str[i] - '0';
    }
    return sum % 10 == 0;
}

CardType get_card_type(long n) {
    char n_str[21];  // LONG_MAX = 9223372036854775807
    sprintf(n_str, "%ld", n);
    CardType type = INVALID;
    char pre[3];
    strncpy(pre, n_str, 2);
    pre[2] = '\0';
    int len = strlen(n_str);
    if (len == 15 && (!strcmp(pre, "34") || !strcmp(pre, "37"))) {
        type = AMEX;
    } else if (len == 16 && strcmp(pre, "50") > 0 && strcmp(pre, "56") < 0) {
        type = MASTERCARD;
    } else if ((len == 13 || len == 16) && pre[0] == '4') {
        type = VISA;
    }
    type = is_valid_number(n_str) ? type : INVALID;
    return type;
}

int main(void) {
    long n = get_long();
    printf("%s\n", CardTypeToString[get_card_type(n)]);
}

File: credit.py
from enum import Enum

class CardType(Enum):
    AMEX = 1
    MASTERCARD = 2
    VISA = 3
    INVALID = 4

def get_card_number():
    while (True):
        try:
            n = int(input("Number: "))
            return n
        except:
            continue

def main():
    n_str = str(get_card_number())
    type = CardType.INVALID
    if len(n_str) == 15 and n_str[:2] in ["34", "37"]:
        type = CardType.AMEX
    elif len(n_str) == 16 and n_str[:2] in ["51", "52", "53", "54", "55"]:
        type = CardType.MASTERCARD
    elif len(n_str) in [13, 16] and n_str[0] == "4":
        type = CardType.VISA
    # Use Luhn's algorithm.
    products = [2*int(i) for i in n_str[-2::-2]]
    total = sum([p if p < 10 else 1 + p%10 for p in products])
    total += sum([int(i) for i in n_str[-1::-2]])
    if total%10 != 0:
        type = CardType.INVALID
    print(type.name)

main()
